File: 0107_Binary_Tree_Level_Order_Traversal_II/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// T = O(n), S = O(n)
class Solution {
  public:
    vector<vector<int>> levelOrderBottom(TreeNode *root) {
        if (root == nullptr) {
            return {};
        }
        vector<vector<int>> res;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            vector<int> v;
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode *p = q.front();
                q.pop();
                v.push_back(p->val);
                if (p->left != nullptr) {
                    q.push(p->left);
                }
                if (p->right != nullptr) {
                    q.push(p->right);
                }
            }
            res.push_back(v);
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: 0107_Binary_Tree_Level_Order_Traversal_II/golang/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrderBottom(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}
	q := []*TreeNode{root}
	res := [][]int{}
	for len(q) != 0 {
		tmp := []*TreeNode{}
		v := []int{}
		for _, p := range q {
			v = append(v, p.Val)
			if p.Left != nil {
				tmp = append(tmp, p.Left)
			}
			if p.Right != nil {
				tmp = append(tmp, p.Right)
			}
		}
		res = append(res, v)
		q = tmp
	}
	n := len(res)
	for i := 0; i < n/2; i++ {
		j := n - i - 1
		res[i], res[j] = res[j], res[i]
	}
	return res
}